src/lib/receiverWithShmem.cc

```diff
@@ -0,0 +1,63 @@
+#include "receiverWithShmem.h"
+
+ReceiverWithShmem::ReceiverWithShmem(std::string shmPath, std::string semProdName, std::string semConsName)
+{
+    this->shmPath = shmPath;
+    this->semProdName = semProdName;
+    this->semConsName = semConsName;
+    this->shmfd = shm_open(shmPath.c_str(), O_RDWR | O_CREAT, 0666);
+    if (shmfd == -1)
+    {
+        throw std::runtime_error("shm_open():" + std::string(strerror(errno)));
+    }
+    this->memptr = mmap(NULL, BUF_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, shmfd, 0);
+    if (memptr == (void *)-1)
+    {
+        throw std::runtime_error("mmap():" + std::string(strerror(errno)));
+    }
+    shm_ctrl = reinterpret_cast<Shmem_control *>(static_cast<char *>(memptr) + BUF_SIZE);
+    semHandleProd = sem_open(semProdName.data(), O_RDWR);
+    // try solving the problem of running receiver earlier that sender here
+    while (semHandleProd == SEM_FAILED)
+    {
+        //    throw std::runtime_error("sem_open() for producer:"+std::string(strerror(errno)));
+        std::cout << "Waiting for the sender to run..." << std::endl;
+        std::cout << "If you have run the sender, sem_open() for producer fails" << std::endl;
+        sleep(2);
+        semHandleProd = sem_open(semProdName.data(), O_RDWR);
+    }
+
+    semHandleCons = sem_open(semConsName.data(), O_RDWR);
+    if (semHandleCons == SEM_FAILED)
+    {
+        throw std::runtime_error("sem_open() for consumer:" + std::string(strerror(errno)));
```

Comment:
If this throw happen, the shared memory won't be munmaped and closed. (when a constructor of a class throws, its own destructor won't be called)